File: cpp/creational/prototype.cpp
/* Prototype Pattern.
 *
 * > Specify the kinds of objects to create using
 * > a prototypical instance, and create new
 * > objects by copying this prototype.
 *
 * This sample uses the pattern on the following
 * context:
 *     - An application share data between
 *         components through a network;
 *     - The actual format and configuration (i.e.,
 *         metadata) for a load
 *         varies greatly and should be changeable
 *         during runtime;
 *     - The payload (i.e., the actual data to be
 *         shared) is always a mapping from keys
 *         to values compatible with JSON;
 */
#include <cmath>
#include <cstdio>
#include <iostream>
#include <memory>
#include <regex>
#include <type_traits>
using namespace std;

// === Utilities ===
struct Runtime_error {
  string message;
};

template <typename T>
typename std::enable_if<std::is_integral<T>::value, string>::type
map_to_str(map<string, T> value) {
  char msg[1024] = "";
  sprintf(msg, "{\n");
  int i = 0, size = value.size();
  for (const auto &tuple : value) {
    sprintf(msg, "%s  \"%s\": %s", msg, tuple.first.c_str(),
            to_string(tuple.second).c_str());
    if (i++ + 1 < size) {
      sprintf(msg, "%s,", msg);
    }
    sprintf(msg, "%s\n", msg);
  }
  sprintf(msg, "%s}\n", msg);
  return msg;
}

string remove_whitespaces(string value) {
  const auto to_remove1 = remove(value.begin(), value.end(), '\n');
  value.erase(to_remove1, value.end());
  const auto to_remove2 = remove(value.begin(), value.end(), ' ');
  value.erase(to_remove2, value.end());
  return value;
}

string caesar_encode(string value, int shift) {
  transform(value.cbegin(), value.cend(), value.begin(),
            [shift](auto ch) { return ch + shift; });
  return value;
}

/*
 * Data (payload) with metadata.
 * */
template <typename T> class LoadPrototype {
  static_assert(is_integral<T>::value, "T must be of integral type.");

public:
  virtual void initialize(map<string, T> payload) = 0;
  virtual void send() = 0;
  virtual unique_ptr<LoadPrototype<T>> clone() = 0;
};

/*
 * Manager of available
 * prototypes for data transmission.
 * */
template <typename T> class LoadManager {
private:
  map<string, shared_ptr<LoadPrototype<T>>> prototypes;
  shared_ptr<LoadPrototype<T>> _default;

public:
  LoadManager() : _default{nullptr} {}

  void add(shared_ptr<LoadPrototype<T>> prototype, string key) {
    this->add(prototype, key, false);
  }

  void add(shared_ptr<LoadPrototype<T>> prototype, string key,
           bool is_default) {
    if (this->prototypes.count(key) > 0) {
      throw Runtime_error{"Key already exists."};
    }
    this->prototypes[key] = prototype;
    if (is_default) {
      this->_default = prototype;
    }
  }

  shared_ptr<LoadPrototype<T>> get(string key) { return this->prototypes[key]; }
};

template <typename T> class PlainLoad : public LoadPrototype<T> {
private:
  bool should_encode = false;
  map<string, T> payload;

public:
  PlainLoad() {}

  PlainLoad(bool should_encode) : should_encode{should_encode} {}

  void initialize(map<string, T> payload) { this->payload = payload; }

  void send() {
    string base = remove_whitespaces(map_to_str(this->payload));
    if (this->should_encode) {
      base = "ENCODED: " + caesar_encode(base, 10);
    }
    printf("[PlainLoad] %s\n", base.c_str());
  }

  unique_ptr<LoadPrototype<T>> clone() {
    auto self = make_unique<PlainLoad<T>>(this->should_encode);
    self->initialize(this->payload);
    return self;
  }
};

template <typename T> class HttpLoad : public LoadPrototype<T> {
private:
  string url;
  map<string, T> payload;

public:
  HttpLoad(string url) : url{url} {}

  void initialize(map<string, T> payload) { this->payload = payload; }

  void send() {
    string base = map_to_str(this->payload);
    printf("[HttpLoad] POST to %s with body:\n%s\n", this->url.c_str(),
           base.c_str());
  }

  unique_ptr<LoadPrototype<T>> clone() {
    auto self = make_unique<HttpLoad<T>>(this->url);
    self->initialize(this->payload);
    return self;
  }
};

/*
 * Sample execution.
 * */
int main() {
  // Initialize RNG
  srand(time(nullptr));

  // Create manager
  LoadManager<int> registry;

  // Register available prototypes
  registry.add(make_shared<PlainLoad<int>>(), "plain");
  registry.add(make_shared<PlainLoad<int>>(true), "encoded");
  registry.add(make_shared<HttpLoad<int>>("http://localhost:8484"), "http");
  string available[3]{"plain", "encoded", "http"};

  // Simulate components
  for (int i = 0; i < 10; i++) {
    // Get prototype
    auto prototype = registry.get(available[random() % 3]);

    // Initialize load
    auto load = prototype->clone();
    load->initialize({{"value", random() % 1001}});

    // Send data
    load->send();
  }

  return 0;
}

File: python/creational/prototype.py
"""Prototype Pattern.

> Specify the kinds of objects to create using
> a prototypical instance, and create new
> objects by copying this prototype.

This sample uses the pattern on the following
context:
    - An application share data between
        components through a network;
    - The actual format and configuration (i.e.,
        metadata) for a load
        varies greatly and should be changeable
        during runtime;
    - The payload (i.e., the actual data to be
        shared) is always a mapping from keys
        to values compatible with JSON;
"""

import base64
import json
import random
from abc import ABC, abstractmethod
from datetime import datetime


class LoadPrototype(ABC):
    """Data (payload) with metadata."""

    @abstractmethod
    def initialize(self, payload: dict):
        """Initialize the load with the
        payload.
        """

    @abstractmethod
    def send(self):
        """Send the payload."""

    @abstractmethod
    def clone(self) -> "LoadPrototype":
        """Return clone of self."""


class LoadManager:
    """Manager of available
    prototypes for data transmission.
    """

    def __init__(self):
        self._prototypes: dict[str, LoadPrototype] = dict()
        self._default: LoadPrototype = None

    def register(self, prototype: LoadPrototype, key: str, is_default: bool = False):
        """Register new prototype.

        :param prototype: prototype.
        :param key: unique identifier.
        :param is_default: if this is the default prototype
            when `get` is called without a key.
        """
        assert key not in self._prototypes
        self._prototypes[key] = prototype

        if is_default:
            self._default = prototype

    def get(self, key: str = None) -> LoadPrototype:
        if key is None:
            assert self._default is not None
            return self._default

        return self._prototypes[key]


class PlainLoad(LoadPrototype):
    def __init__(self, should_encode: bool = False, payload: dict = None):
        self._should_encode = should_encode
        self._payload = payload

    def initialize(self, payload: dict):
        self._payload = payload

    def send(self):
        assert self._payload is not None
        data = str(self._payload)
        if self._should_encode:
            data = "ENCODED: " + base64.b64encode(data.encode()).decode()
        print(f"[PlainLoad] {data}")

    def clone(self) -> "PlainLoad":
        return PlainLoad(self._should_encode, self._payload)


class HttpLoad(LoadPrototype):
    def __init__(self, url: str, payload: dict = None):
        self._url = url
        self._payload = payload

    def initialize(self, payload: dict):
        self._payload = payload

    def send(self):
        assert self._payload is not None
        data = dict(
            timestamp=datetime.now().isoformat(), version=1.0, data=self._payload
        )
        print(f"[HttpLoad] POST to {self._url} with body:")
        print(json.dumps(data, indent=2, ensure_ascii=False))

    def clone(self) -> "HttpLoad":
        return HttpLoad(self._url, self._payload)


if __name__ == "__main__":
    # Create manager
    registry = LoadManager()

    # Register available prototypes
    registry.register(PlainLoad(should_encode=False), key="plain", is_default=True)
    registry.register(PlainLoad(should_encode=True), key="base64")
    registry.register(HttpLoad("http://localhost:8484"), key="http")

    # Simulate components
    for _ in range(10):
        # Generate random payload
        payload = dict(value=random.randint(0, 1000))

        # Get prototype
        prototype = registry.get(random.choice([None, "plain", "base64", "http"]))

        # Initialize load
        load = prototype.clone()
        load.initialize(payload)

        # Send data
        load.send()
